lib/gateway.js

```diff
@@ -0,0 +1,236 @@
+'use strict'
+
+const buildGateway = require('./gateway/build-gateway')
+const {
+  MER_ERR_INVALID_OPTS,
+  MER_ERR_GQL_GATEWAY,
+  MER_ERR_GQL_GATEWAY_INIT
+} = require('./errors')
+const { kHooks } = require('mercurius/lib/symbols')
+const {
+  onGatewayReplaceSchemaHandler,
+  assignApplicationLifecycleHooksToContext
+} = require('./handlers')
+const { buildCache } = require('./util')
+
+function validateGateway(opts) {
+  const gateway = opts
+
+  if (Array.isArray(gateway.services)) {
+    const serviceNames = new Set()
+    for (const service of gateway.services) {
+      if (typeof service !== 'object') {
+        throw new MER_ERR_INVALID_OPTS(
+          'gateway: all "services" must be objects'
+        )
+      }
+      if (typeof service.name !== 'string') {
+        throw new MER_ERR_INVALID_OPTS(
+          'gateway: all "services" must have a "name" String property'
+        )
+      }
+      if (serviceNames.has(service.name)) {
+        throw new MER_ERR_INVALID_OPTS(
+          `gateway: all "services" must have a unique "name": "${service.name}" is already used`
+        )
+      }
+      serviceNames.add(service.name)
+      if (
+        typeof service.url !== 'string' &&
+        (!Array.isArray(service.url) ||
+          service.url.length === 0 ||
+          !service.url.every(url => typeof url === 'string'))
+      ) {
+        throw new MER_ERR_INVALID_OPTS(
+          'gateway: all "services" must have an "url" String, or a non-empty Array of String, property'
+        )
+      }
+    }
+  } else {
+    throw new MER_ERR_GQL_GATEWAY_INIT(
+      'The "services" attribute cannot be undefined'
+    )
+  }
+}
+
+async function createGateway(gatewayOpts, app) {
+  validateGateway(gatewayOpts)
+
+  const retryServicesCount =
+    gatewayOpts && gatewayOpts.retryServicesCount
+      ? gatewayOpts.retryServicesCount
+      : 10
+
+  const retryInterval = gatewayOpts.retryServicesInterval || 3000
+
+  const lruGatewayResolvers = buildCache(gatewayOpts)
+
+  const serviceMap = {}
+
+  try {
+    const gateway = await buildGateway(
+      serviceMap,
+      gatewayOpts,
+      app,
+      lruGatewayResolvers
+    )
+
+    app.graphql.replaceSchema(gateway.schema)
+    // let isModuleInitialized = false
+    // app.addHook('onRegister', () => {
```

Comment:
Thanks, I'll remove it.
Yes, it needs a clean up. I'm addressing some comments and this is still in progress. 

I've reconverted the PR in `draft` to avoid it.